### PR TITLE
[SPARK-15782] [YARN] Set spark.jars system property in client mode

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
## What changes were proposed in this pull request?

When `--packages` is specified with `spark-shell` the classes from those packages cannot be found, which I think is due to some of the changes in `SPARK-12343`. In particular `SPARK-12343` removes a line that sets the `spark.jars` system property in client mode, which is used by the repl main class to set the classpath.
## How was this patch tested?

Tested manually.

This system property is used by the repl to populate its classpath. If
this is not set properly the classes for external packages cannot be
found.

@tgravescs @vanzin as you may be familiar with this part of the code.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -444,6 +444,7 @@ object SparkSubmit {
       OptionAssigner(args.deployMode, ALL_CLUSTER_MGRS, ALL_DEPLOY_MODES,
         sysProp = "spark.submit.deployMode"),
       OptionAssigner(args.name, ALL_CLUSTER_MGRS, ALL_DEPLOY_MODES, sysProp = "spark.app.name"),
+      OptionAssigner(args.jars, ALL_CLUSTER_MGRS, CLIENT, sysProp = "spark.jars"),
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: I'm pretty sure this is not what you want to do. If you look later, you have these lines:

```
OptionAssigner(args.jars, YARN, ALL_DEPLOY_MODES, sysProp = "spark.yarn.dist.jars"),
OptionAssigner(args.jars, LOCAL, CLIENT, sysProp = "spark.jars"),
OptionAssigner(args.jars, STANDALONE | MESOS, ALL_DEPLOY_MODES, sysProp = "spark.jars"),
```

So I see how the change caused an issue with the repl in yarn-client mode. But your fix would cause issues with the yarn backend, where the same jars would be distributed twice (via "spark.jars" and "spark.yarn.dist.jars").

But it seems like the issue you found is not restricted to the shell itself; SparkContext seems to also only look at `spark.jars` and not the yarn-specific one. So jars specified in the command line would not show up in the driver's class loader in yarn-client mode.

I see two options here:
- revert the original change that used the distributed cache in yarn-client mode; this seems like the safer bet for Spark 2. That would mean adding this line you're adding, and also removing the others.
- make SparkContext and other places that only process `spark.jars` also process `spark.yarn.dist.jars`. This seems like something that we have to do to properly support `spark.yarn.dist.jars` anyway.

/cc @jerryshao since he made the original changes.

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: From my understanding, I think option 2 might be a good choice.

Comment (nezihyigitbasi) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Thanks @vanzin @jerryshao for your input, some follow up questions:
- How should Spark behave when a user both sets `spark.jars` and `spark.yarn.dist.jars`? Do we want to set union them? (the comment on this method says "Set JAR files to distribute to the cluster")
- In `SparkConf` there is a `setJars()` method that sets `spark.jars`, should this method also set `spark.yarn.dist.jars` if we are in YARN mode?
- Does implementing option 2 resolve the issue completely? Because the repl main class needs the `spark.jar` system property to be set. Shouldn't we also set that system property somewhere?

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: > How should Spark behave when a user both sets spark.jars and spark.yarn.dist.jars? Do we want to set union them?

Yes, I think that should be the expected behavior. There might be some weird semantics that need to be addressed, though:

- user sets `spark.jars` in conf
- SparkSubmitArguments sets `args.jars` to value of `spark.jars`
- YARN-specific code sets `spark.yarn.dist.jars` to value of `args.jars`
- Now both `spark.jars` and `spark.yarn.dist.jars` have the same value.

Perhaps when the YARN translation happens above, `spark.jars` should be explicitly unset.

> In SparkConf there is a setJars() method that sets spark.jars

I think it's ok to leave that as is. It could be done as an optimization, but it's not necessarily related to this bug so let's leave that for a different change, if desired.

> Does implementing option 2 resolve the issue completely?

I did some quick testing and it seems to work. You just need to make sure the code works regardless of which option is set. The classpath for the repl should be the concatenation of both options.

Optimally the repl would only look at the YARN-specific options when running in YARN mode, but I think that code runs too early in the process for Spark to already know whether it will use YARN or not.

Also remember you need to change both the 2.10 and 2.11 repl code. Might be easier if you make it a helper method in `Utils.scala`, for example.

Comment (nezihyigitbasi) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: thanks @vanzin, will start working on this shortly.

**Comment by nezihyigitbasi** [comment]
@vanzin updated to address your comments.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -391,7 +391,9 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
     _conf.set("spark.executor.id", SparkContext.DRIVER_IDENTIFIER)
 
-    _jars = _conf.getOption("spark.jars").map(_.split(",")).map(_.filter(_.nonEmpty)).toSeq.flatten
+    val sparkJars = _conf.getOption("spark.jars").getOrElse("")
+    val yarnDistJars = _conf.getOption("spark.yarn.dist.jars").getOrElse("")
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Should only do this if `master` == `yarn`.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 3]
```diff
@@ -484,6 +484,13 @@ object SparkSubmit {
       OptionAssigner(args.ivyRepoPath, STANDALONE, CLUSTER, sysProp = "spark.jars.ivy")
     )
 
+    // In YARN mode if we have the yarn dist jars and spark jars have the same value
+    // then we can unset spark jars
+    if (clusterManager == YARN &&
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Is this necessary? Won't the code in SparkContext already take care of this?

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 4]
```diff
@@ -2305,6 +2305,23 @@ private[spark] object Utils extends Logging {
     log.info(s"Started daemon with process name: ${Utils.getProcessName()}")
     SignalUtils.registerLogger(log)
   }
+
+  /**
+  * Unions two comma-separated lists of files.
+  */
+  def setUnionFileLists(leftList: String, rightList: String): Set[String] = {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: This is a weird name. How about just `unionFileLists`? You could also use varargs for the arguments instead of just allowing two lists, but that's minor.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 5]
```diff
@@ -2305,6 +2305,23 @@ private[spark] object Utils extends Logging {
...
+    val leftSet = if (leftList == "" || leftList == null) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: If you use `Option[String]` as the argument you can avoid this and the `getOrElse` calls in SparkCOntext, and do something like:

```
var allFiles = Set[String]()

list.foreach { value => allFiles ++= value.split(",") }
```

Which is a lot shorter than the code you have.

**Comment by nezihyigitbasi** [comment]
thanks @vanzin, addressed your comments.

**Comment by vanzin** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 6]
```diff
@@ -2305,6 +2305,16 @@ private[spark] object Utils extends Logging {
...
   }
+
+  /**
+  * Unions two comma-separated lists of files and filters out empty strings.
+  */
+  def unionFileLists(leftList: Option[String], rightList: Option[String]): Set[String] = {
+    var allFiles = Set[String]()
+    leftList.foreach { value => allFiles ++= value.split(",") }
+    rightList.foreach { value => allFiles ++= value.split(",") }
+    allFiles.filter { value => value != "" }
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: `allFiles.filter(_.nonEmpty)`

File: repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala [hunk 7]
```diff
@@ -1071,8 +1071,16 @@ object SparkILoop extends Logging {
     if (envJars.isDefined) {
```
Comment (vanzin) [inline repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala]: If you want to do something else, you could file a separate bug to remove this. Not only this is deprecated but it doesn't even exist in the 2.11 version, so might as well get rid of it.

**Comment by vanzin** [comment]
LGTM pending tests. The 2.10 shell code could be cleaned up, but that would be better done as part of removing `ADD_JARS`.

File: repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala (hunk 7 above)
Comment (nezihyigitbasi) [inline repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkILoop.scala]: let's keep it here, we can do it in a separate PR.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 8]
```diff
@@ -2305,6 +2305,16 @@ private[spark] object Utils extends Logging {
     log.info(s"Started daemon with process name: ${Utils.getProcessName()}")
     SignalUtils.registerLogger(log)
   }
+
+  /**
+  * Unions two comma-separated lists of files and filters out empty strings.
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: nit: alignment is off (needs an extra space)

File: repl/scala-2.11/src/main/scala/org/apache/spark/repl/Main.scala [hunk 9]
```diff
@@ -54,9 +54,9 @@ object Main extends Logging {
   // Visible for testing
   private[repl] def doMain(args: Array[String], _interp: SparkILoop): Unit = {
     interp = _interp
-    val jars = conf.getOption("spark.jars")
-      .map(_.replace(",", File.pathSeparator))
-      .getOrElse("")
+    val sparkJars = conf.getOption("spark.jars")
+    val yarnJars = conf.getOption("spark.yarn.dist.jars")
```
Comment (vanzin) [inline repl/scala-2.11/src/main/scala/org/apache/spark/repl/Main.scala]: Similar to the `SparkContext` check, this probably should only be done is `conf.get("spark.master")` is "yarn".

**Comment by vanzin** [comment]
@nezihyigitbasi I went through the change again and I think the code should be a little more careful about only looking at the YARN settings if actually running on YARN.

You could also make that easier by having a method in Utils to return the list of user jars given a `SparkConf` object. Then you could just call that method from the three different places, instead of the current code which looks very similar in its 3 versions.

**Comment by nezihyigitbasi** [comment]
@vanzin makes sense, updated.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 10]
```diff
@@ -39,7 +39,7 @@ import scala.util.control.{ControlThrowable, NonFatal}
 
 import com.google.common.io.{ByteStreams, Files => GFiles}
 import com.google.common.net.InetAddresses
-import org.apache.commons.lang3.SystemUtils
+import org.apache.commons.lang3.{StringUtils, SystemUtils}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: nit: this change is not needed now

**Comment by vanzin** [comment]
LGTM bar one tiny nit. Also, the title should probably say "yarn" instead of "core" since this only really affects YARN mode.

**Comment by nezihyigitbasi** [comment]
@vanzin updated, thanks a lot for reviewing this one.

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
Failures are unrelated (and one run passed). Merging to master / 2.0.

**Comment by rxin** [comment]
This broke 2.10. Can somebody do a quick fix?

**Comment by vanzin** [comment]
Let me take a look.
